File: ch4/python/code4_8.py

def max_sub_array_sum(arr, n):
    # C 예제의 INT_MIN과 같이 충분히 작은 값을 파이썬에서는
    # -float('inf')로 지정해 사용할 수 있습니다.
    max_sum = -float('inf')
    
    for i in range(0, n):
        temp_sum = 0
        for j in range(i, n):
            temp_sum += arr[j]
            if temp_sum > max_sum:
                max_sum = temp_sum
    return max_sum

arr = [-2, -3, 4, -1, -2, 1, 5, -3]
n = len(arr)
print('부분 배열의 합의 최댓값은 %d입니다.' % max_sub_array_sum(arr, n))

File: ch4/c/code4_8.c
#include <stdio.h>
#include <limits.h>

int maxSubArraySum(int* arr, int n)
{
  int maxSum = INT_MIN;
  int tempSum = 0;

  for(int i = 0; i < n; i++)
  {
    tempSum = 0;
    for(int j = i; j < n; j++)
    {
      // tempSum은 현재의 인덱스 i에서 인덱스 j까지의 배열의
      // 숫자의 합을 저장합니다.
      tempSum += arr[j];
      if(tempSum > maxSum)
        maxSum = tempSum;
    }
  }
  return maxSum;
}

int main()
{
  int arr[] = {-2, -3, 4, -1, -2, 1, 5, -3};
  int n = sizeof(arr) / sizeof(int);
  printf("부분 배열의 합의 최대값은 %d입니다.\n", maxSubArraySum(arr, n));
  return 0;
}